runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // First generate the OOL code. This path will be used to handle the scenarios that we don't want to handle in JITT'ed code.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);
+
+   TR::Node *_callNode = createOutlinedCallNode(cg, node, TR::acall);
+   TR::Register *targetReg2 = NULL;
+   targetReg2 = TR::TreeEvaluator::performCall(_callNode, false, cg);
+   cg->decReferenceCount(_callNode);
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, targetReg, targetReg2);
+
+   // After returning from the helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   // End of OOL
+   outlinedSlowPath->swapInstructionListsWithCompilation();
 
+   // In the mainline, first load the first and second dimensions' lengths into registers.
+   TR::Register *firstDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, firstDimLenReg, generateS390MemoryReference(dimsPtrReg, 4, cg));
+
+   TR::Register *secondDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, secondDimLenReg, generateS390MemoryReference(dimsPtrReg, 0, cg));
+
+
+   // Check to see if second dimension is indeed 0. If yes, then proceed to handle the case here. Otherwise jump to OOL code to perform the call.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, secondDimLenReg, (int8_t)0, oolFailLabel, (TR::InstOpCode::COND_BNE));
+
+   // Now check to see if first dimension is also 0. If no, then jump to nonZeroFirstDimLabel to handle it. Otherwise continue below to handle the case when length for both dimensions is 0.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, firstDimLenReg, (int8_t)0, nonZeroFirstDimLabel, (TR::InstOpCode::COND_BNE));
+
+   // TODO: We don't need this label. But it's good for organizing the different paths for now.
+   TR::LabelSymbol *firstDimZeroLabel = generateLabelSymbol(cg);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, firstDimZeroLabel);
```

Comment:
I think this clouds the evaluator. When generating a trace log if I see a label I'd expect there is an incoming branch somewhere. So this will definitely confuse readers. If you want to separate code and organize it, just use a comment.